proto/document/v1/document.proto

```diff
@@ -0,0 +1,104 @@
+syntax = "proto3";
+package nitric.document.v1;
+
+import "google/protobuf/struct.proto";
+
+// protoc plugin options for code generation
+option go_package = "nitric/v1;v1";
+option java_package = "io.nitric.proto.document.v1";
+option java_multiple_files = true;
+option java_outer_classname = "Documents";
+option php_namespace = "Nitric\\Proto\\Document\\V1";
+option csharp_namespace = "Nitric.Proto.Document.v1";
+
+// Service for storage and retrieval of simple JSON keyValue
+service Document {
+  // Get an existing key
+  rpc Get (DocGetRequest) returns (DocGetResponse);
+
+  // Create a new or overwrite and existing key
+  rpc Put (DocPutRequest) returns (DocPutResponse);
+
+  // Delete an existing
+  rpc Delete (DocDeleteRequest) returns (DocDeleteResponse);
+  
+  // Query the collection
+  rpc Query (DocQueryRequest) returns (DocQueryResponse);
+}
+
+message Key {
+  // The item collection
+  string collection = 1;
+
+  // The items unique id
+  string id = 2;
+}
+
+message DocGetRequest {
+  // The parent root key
+  Key key = 1;
+
+  // The optional sub collection key
+  Key subKey = 2;
+}
+
+message DocGetResponse {
+  // The retrieved value
+  google.protobuf.Struct value = 1;
+}
+
+message DocPutRequest {
+  // The parent root key
+  Key key = 1;
+
+  // The optional sub collection key
+  Key subKey = 2;
+
+  // A simple JSON object to store
+  google.protobuf.Struct value = 3;
+}
+
+message DocPutResponse {}
+
+message DocDeleteRequest {
+  // The parent root key
+  Key key = 1;
+
+  // The sub collection key (optional)
+  Key subKey = 2;
+}
+
+message DocDeleteResponse {}
+
+message DocQueryRequest {
```

Comment:
For sub collection queries we could potentially include a new request type that include the standard DocQueryRequest inside it.

```
message DocSubCollectionQueryRequest {
  // Adding this parent key should give enough context to scope the subcollection query
  Key parentKey = 1;
  
  DocQueryRequest query = 2;
```

Reply:
Or even include the parentKey as an optional parameter in the DocQueryRequest, which when not null means it's a subcollection query

Reply:
that looks interesting